File: src/handlers/frontmatter.rs
use crate::error::{ErrorType, MapProcErr, ProcessError, WithItem};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Deserialize, Serialize)]
pub struct FileList {
    pub files: Vec<String>,
}

/// Result of frontmatter extraction: the key-value map, remaining content, and any warnings.
pub struct FrontmatterResult {
    pub map: HashMap<String, String>,
    pub remaining: String,
    pub warnings: Vec<ProcessError>,
}

/// Extract YAML frontmatter from markdown content.
/// Returns the frontmatter map, remaining content, and any non-fatal warnings
/// (e.g. unsupported value types that were skipped).
pub fn extract_frontmatter(
    content: &str,
    path: &PathBuf,
) -> Result<FrontmatterResult, ProcessError> {
    let content = content.trim_start();

    if !content.starts_with("---") {
        return Err(ProcessError {
            error_type: ErrorType::Syntax,
            item: WithItem::Data,
            path: path.clone(),
            message: Some("Frontmatter must start with '---'".to_string()),
        });
    }

    let after_first_delimiter = &content[3..];

    if let Some(end_pos) = after_first_delimiter.find("\n---") {
        let frontmatter_str = &after_first_delimiter[..end_pos];
        let remaining = &after_first_delimiter[end_pos + 4..].trim_start();

        // Parse YAML frontmatter
        let yaml_value: serde_yaml::Value =
            serde_yaml::from_str(frontmatter_str).map_err(|e| ProcessError {
                error_type: ErrorType::Syntax,
                item: WithItem::Data,
                path: path.clone(),
                message: Some(format!("Failed to parse YAML frontmatter: {}", e)),
            })?;

        let mut map = HashMap::new();
        let mut warnings = Vec::new();

        if let serde_yaml::Value::Mapping(mapping) = yaml_value {
            for (key, value) in mapping {
                if let serde_yaml::Value::String(k) = &key {
                    let v = match &value {
                        serde_yaml::Value::String(s) => s.clone(),
                        serde_yaml::Value::Number(n) => n.to_string(),
                        serde_yaml::Value::Bool(b) => b.to_string(),
                        serde_yaml::Value::Null => {
                            warnings.push(ProcessError {
                                error_type: ErrorType::Syntax,
                                item: WithItem::Data,
                                path: path.clone(),
                                message: Some(format!(
                                    "Frontmatter key '{}' has a null value and was skipped",
                                    k
                                )),
                            });
                            continue;
                        }
                        other => {
                            let type_name = match other {
                                serde_yaml::Value::Sequence(_) => "array",
                                serde_yaml::Value::Mapping(_) => "nested mapping",
                                serde_yaml::Value::Tagged(_) => "tagged value",
                                _ => "unsupported type",
                            };
                            warnings.push(ProcessError {
                                error_type: ErrorType::Syntax,
                                item: WithItem::Data,
                                path: path.clone(),
                                message: Some(format!(
                                    "Frontmatter key '{}' has an unsupported type ({}) and was skipped. \
                                     Only strings, numbers, and booleans are supported.",
                                    k, type_name
                                )),
                            });
                            continue;
                        }
                    };
                    map.insert(k.clone(), v);
                }
            }
        }

        // Validate that title exists
        if !map.contains_key("title") {
            return Err(ProcessError {
                error_type: ErrorType::Syntax,
                item: WithItem::Data,
                path: path.clone(),
                message: Some("Frontmatter must contain a 'title' field".to_string()),
            });
        }

        Ok(FrontmatterResult {
            map,
            remaining: remaining.to_string(),
            warnings,
        })
    } else {
        Err(ProcessError {
            error_type: ErrorType::Syntax,
            item: WithItem::Data,
            path: path.clone(),
            message: Some("Frontmatter must end with '---'".to_string()),
        })
    }
}

/// Load data from markdown files with frontmatter based on a TOML file list.
/// Returns a JSON array value compatible with the existing template system.
pub fn load_frontmatter_data(
    src: &PathBuf,
    name: &str,
) -> Result<(Value, Vec<ProcessError>), Vec<ProcessError>> {
    let mut errors = Vec::new();

    let toml_path = src
        .join("data")
        .join(name.replace(":", "/"))
        .with_extension("data.toml");

    // Read the TOML file
    let toml_content = fs::read_to_string(&toml_path)
        .map_proc_err(
            WithItem::Data,
            ErrorType::Io,
            &toml_path,
            Some("Failed to read data.toml file".to_string()),
        )
        .map_err(|e| vec![e])?;

    // Parse the TOML file
    let file_list: FileList = toml::from_str(&toml_content).map_err(|e| {
        vec![ProcessError {
            error_type: ErrorType::Syntax,
            item: WithItem::Data,
            path: toml_path.clone(),
            message: Some(format!("Failed to parse TOML: {}", e)),
        }]
    })?;

    let data_dir = src.join("data").join(name.replace(":", "/"));
    let mut items = Vec::new();

    for file in &file_list.files {
        let md_path = data_dir.join(file);

        let content = match fs::read_to_string(&md_path) {
            Ok(c) => c,
            Err(e) => {
                errors.push(ProcessError {
                    error_type: ErrorType::Io,
                    item: WithItem::Data,
                    path: md_path.clone(),
                    message: Some(format!("Failed to read markdown file: {}", e)),
                });
                continue;
            }
        };

        let fm_result = match extract_frontmatter(&content, &md_path) {
            Ok(r) => r,
            Err(e) => {
                errors.push(e);
                continue;
            }
        };

        // Collect frontmatter warnings
        errors.extend(fm_result.warnings);

        let mut frontmatter = fm_result.map;

        // Generate entry-path and result-path from filename
        let file_stem = md_path
            .file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("unknown");

        let relative_entry_path = format!("{}/{}", name.replace(":", "/"), file);
        let result_path = format!("content/{}.html", file_stem);

        // Add the special fields
        frontmatter.insert("--entry-path".to_string(), relative_entry_path);
        frontmatter.insert("--result-path".to_string(), result_path.clone());
        frontmatter.insert("link".to_string(), format!("./{}", result_path));

        // Convert HashMap to serde_json::Value
        let obj: serde_json::Map<String, Value> = frontmatter
            .into_iter()
            .map(|(k, v)| (k, Value::String(v)))
            .collect();

        items.push(Value::Object(obj));
    }

    Ok((Value::Array(items), errors))
}

File: src/handlers/frontmatter.cpp
#include "handlers/frontmatter.hpp"
#include <format>
#include <fstream>
#include <sstream>
#include <toml++/toml.hpp>
#include <yaml-cpp/yaml.h>

namespace simple::handlers {

auto extract_frontmatter(std::string_view content) -> Result<
    std::pair<std::unordered_map<std::string, std::string>, std::string>> {

  size_t pos = 0;
  while (pos < content.size() && std::isspace(content[pos])) {
    ++pos;
  }
  content.remove_prefix(pos);

  if (!content.starts_with("---")) {
    return std::unexpected(ProcessError{ErrorType::Syntax,
                                        WithItem::Data,
                                        {},
                                        "Frontmatter must start with '---'"});
  }

  auto after_first = content.substr(3);
  auto end_pos = after_first.find("\n---");

  if (end_pos == std::string_view::npos) {
    return std::unexpected(ProcessError{ErrorType::Syntax,
                                        WithItem::Data,
                                        {},
                                        "Frontmatter must end with '---'"});
  }

  auto frontmatter_str = after_first.substr(0, end_pos);
  auto remaining = after_first.substr(end_pos + 4);

  while (!remaining.empty() && std::isspace(remaining[0])) {
    remaining.remove_prefix(1);
  }

  try {
    auto yaml = YAML::Load(std::string{frontmatter_str});
    std::unordered_map<std::string, std::string> map;

    if (yaml.IsMap()) {
      for (const auto &kv : yaml) {
        auto key = kv.first.as<std::string>();
        std::string value;

        if (kv.second.IsScalar()) {
          value = kv.second.as<std::string>();
        } else {
          value = "";
        }

        map[key] = value;
      }
    }

    if (!map.contains("title")) {
      return std::unexpected(
          ProcessError{ErrorType::Syntax,
                       WithItem::Data,
                       {},
                       "Frontmatter must contain a 'title' field"});
    }

    return std::make_pair(std::move(map), std::string{remaining});

  } catch (const YAML::Exception &e) {
    return std::unexpected(ProcessError{
        ErrorType::Syntax,
        WithItem::Data,
        {},
        std::format("Failed to parse YAML frontmatter: {}", e.what())});
  }
}

auto load_frontmatter_data(const std::filesystem::path &src,
                           std::string_view name)
    -> std::expected<std::pair<nlohmann::json, std::vector<ProcessError>>,
                     std::vector<ProcessError>> {

  std::vector<ProcessError> errors;

  std::string name_str{name};
  size_t pos = 0;
  while ((pos = name_str.find(':', pos)) != std::string::npos) {
    name_str.replace(pos, 1, "/");
  }

  auto toml_path = src / "data" / (name_str + ".data.toml");

  toml::table tbl;
  try {
    tbl = toml::parse_file(toml_path.string());
  } catch (const toml::parse_error &e) {
    return std::unexpected(std::vector{
        ProcessError{ErrorType::Syntax, WithItem::Data, toml_path,
                     std::format("Failed to parse TOML: {}", e.what())}});
  }

  nlohmann::json items = nlohmann::json::array();
  auto data_dir = src / "data" / name_str;

  auto files_node = tbl["files"];
  if (!files_node.is_array()) {
    return std::unexpected(
        std::vector{ProcessError{ErrorType::Syntax, WithItem::Data, toml_path,
                                 "Missing 'files' array in TOML"}});
  }

  auto files_array = *files_node.as_array();
  std::vector<std::string> files;
  for (const auto &item : files_array) {
    if (item.is_string()) {
      files.push_back(std::string{*item.value<std::string>()});
    }
  }

  for (const auto &file : files) {
    auto md_path = data_dir / file;

    std::ifstream md_file{md_path};
    if (!md_file) {
      errors.push_back(
          ProcessError{ErrorType::Io, WithItem::Data, md_path,
                       std::format("Failed to read markdown file: {}", file)});
      continue;
    }

    std::stringstream md_buffer;
    md_buffer << md_file.rdbuf();
    auto content = md_buffer.str();

    auto fm_result = extract_frontmatter(content);
    if (!fm_result) {
      auto err = fm_result.error();
      err.path = md_path;
      errors.push_back(err);
      continue;
    }

    auto [frontmatter, remaining] = std::move(*fm_result);

    auto file_stem = md_path.stem().string();
    auto relative_entry_path = std::format("{}/{}", name_str, file);
    auto result_path = std::format("content/{}.html", file_stem);

    frontmatter["--entry-path"] = relative_entry_path;
    frontmatter["--result-path"] = result_path;
    frontmatter["link"] = std::format("./{}", result_path);

    nlohmann::json obj = nlohmann::json::object();
    for (const auto &[k, v] : frontmatter) {
      obj[k] = v;
    }

    items.push_back(obj);
  }

  return std::make_pair(std::move(items), std::move(errors));
}

} // namespace simple::handlers
